Allow spaces in parser

Matth-- opened the pull request:
The from and to in diffs did not include the full path name when
spaces are being used. This change allows spaces in the file paths
as well (no tabs).

SpacePossum commented:
Hi and nice find!

Just wondering if this can be tweaked even more to something like; `(^---\h+(?P<file>[^\v]+))` as after `---` (or `+++`) there can be any horizontal space (not 100% this must a single space in all the diff formats) and than whatever follows up a vertical line break or end of the string makes up the file name I think.

Matth-- commented:
Hi @SpacePossum 

Good suggestions! I changed the commit. 

I also added tabs to the exclusion list. [This test](https://github.com/sebastianbergmann/diff/blob/master/tests/ParserTest.php#L164) also has timestamps after the filenames (with a leading tab).

If tabs are being used for filenames (why? 😄 ) it gets represented as `\t` and the filenames are surrounded with double-quotes.

```
diff --git "a/\ttabbed.txt" "b/\ttabbed2.txt"
index 9daeafb..e69de29 100644
--- "a/\ttabbed.txt"
+++ "b/\ttabbed2.txt"
@@ -1 +0,0 @@
-test
```

Would this cover everything?

SpacePossum commented:
Good catch on the timestamps! I really wished "they" formalized the diff file format in some standard :D 
It looks good me!

Matth-- commented:
Any idea why the checks failed?

SpacePossum commented:
```diff
diff --git a/src/Parser.php b/src/Parser.php
index 6090e81..cc9e388 100644
--- a/src/Parser.php
+++ b/src/Parser.php
@@ -37,8 +37,8 @@ public function parse(string $string): array
         $collected = [];
 
         for ($i = 0; $i < $lineCount; ++$i) {
-            if (preg_match('(^---\\s+(?P<file>\\S+))', $lines[$i], $fromMatch) &&
-                preg_match('(^\\+\\+\\+\\s+(?P<file>\\S+))', $lines[$i + 1], $toMatch)) {
+            if (preg_match('#^---\h+"?(?P<file>[^\\v\\t"]+)#', $lines[$i], $fromMatch) &&
+                preg_match('#^\\+\\+\\+\\h+"?(?P<file>[^\\v\\t"]+)#', $lines[$i + 1], $toMatch)) {
                 if ($diff !== null) {
                     $this->parseFileDiff($diff, $collected);
 
diff --git a/tests/ParserTest.php b/tests/ParserTest.php
index 7623970..30d9fe2 100644
--- a/tests/ParserTest.php
+++ b/tests/ParserTest.php
@@ -166,4 +166,45 @@ public function diffProvider(): array
             ],
         ];
     }
+
+    public function testParseWithSpacesInFileNames(): void
+    {
+        $content =
+<<<PATCH
+diff --git a/Foo Bar.txt b/Foo Bar.txt
+index abcdefg..abcdefh 100644
+--- a/Foo Bar.txt
++++ b/Foo Bar.txt
+@@ -20,4 +20,5 @@ class Foo
+     const ONE = 1;
+     const TWO = 2;
++    const THREE = 3;
+     const FOUR = 4;
+PATCH;
+
+        $diffs = $this->parser->parse($content);
+
+        $this->assertEquals('a/Foo Bar.txt', $diffs[0]->getFrom());
+        $this->assertEquals('b/Foo Bar.txt', $diffs[0]->getTo());
+    }
+
+    public function testParseWithSpacesInFileNamesAndTimestamp(): void
+    {
+        $content =
+            <<<PATCH
+diff --git a/Foo Bar.txt b/Foo Bar.txt
+index abcdefg..abcdefh 100644
+--- "a b.txt"  2020-10-02 13:31:52.938811371 +0200
++++ "a c.txt"  2020-10-02 13:31:50.022792064 +0200
+@@ -20,4 +20,5 @@ class Foo
+     const ONE = 1;
+     const TWO = 2;
++    const THREE = 3;
+     const FOUR = 4;
+PATCH;
+        $diffs = $this->parser->parse($content);
+
+        $this->assertEquals('a b.txt', $diffs[0]->getFrom());
+        $this->assertEquals('a c.txt', $diffs[0]->getTo());
+    }
 }
```

Hi again!

Above will allow for `--- "a b.txt"	2020-10-02 13:31:52.938811371 +0200` as well (currently incorrectly returns the file name with the quotes attached). It also makes your fixture content inline so we don't need the file itself.
Let me know what you think!

About the test failing, not sure as it doesn't seem related to your changes. Maybe try to kick them again by committing something to the PR as the composer dependencies might be stuck in the cache.

Matth-- commented:
@SpacePossum Good catch!

I implemented your changes and the tests pass.

Do you know how I get that diff output (with timestamps and surrounding quotes)?

SpacePossum commented:
I made the diff examples using `git diff` on my setup.

It is actually a bug that the quotes are not added by https://github.com/sebastianbergmann/diff/blob/4.0.3/src/Output/StrictUnifiedDiffOutputBuilder.php , I'll put that on my todo list (it does allow for adding the timestamps).

The test failures are confusing to me, maybe Sebastian can help out with the setup when he has some time.

SpacePossum commented:
Ping @sebastianbergmann whenever you've got time, please let us know how we can help out, thanks! :) 
This PR looks good to me, but the test setup seems broken.

sebastianbergmann commented:
> This PR looks good to me, but the test setup seems broken.

Maybe @localheinz has some insight why GitHub Actions (sometimes) fail for pull requests on my projects.

@SpacePossum If this is looks okay to you, then it is okay for me as well. I'll merge this which should (fingers crossed) give us a useful GitHub Actions build.

SpacePossum commented:
thanks so much :)
and :wave: to Andreas

thanks for the PR @Matth--  :+1:

localheinz commented:
@sebastianbergmann 

Could this be the composer root version problem again?

sebastianbergmann commented:
Could be, don't know. Sorry!
